torch/csrc/Exceptions.h

```diff
@@ -271,13 +261,50 @@ struct PyWarningHandler: at::WarningHandler {
   void process(const at::SourceLocation &source_location,
                const std::string &msg) override;
 
+  /** Call if an exception has been thrown
+
+   *  Necessary to determine if it is safe to throw from the desctructor since
+   *  std::uncaught_exception is buggy on some platforms and generally
+   *  unreliable across dynamic library calls.
+   */
+  void set_in_exception() {
+    in_exception_ = true;
+  }
+
 private:
   using warning_buffer_t =
     std::vector<std::pair<c10::SourceLocation, std::string>>;
 
   warning_buffer_t warning_buffer_;
 
   at::WarningHandler* prev_handler_;
+  bool in_exception_;
 };
 
+namespace detail {
+template <typename Func, size_t i>
+using Arg = typename function_traits<Func>::template arg<i>::type;
+
+template <typename Func, size_t ...Is>
+auto wrap_pybind_function_impl_(Func&& f, std::index_sequence<Is...>) {
+  using traits = function_traits<Func>;
+  namespace py = pybind11;
+
+  return [f](Arg<Func, Is> ...args) -> typename traits::result_type {
+    HANDLE_TH_ERRORS
+    return f(std::forward<Arg<Func, Is>>(args)...);
+    END_HANDLE_TH_ERRORS_PYBIND
+  };
+}
+}  // namespace detail
+
+// Wrap a function with TH error and warning handling.
+// Returns a function object suitable for registering with pybind11.
+template <typename Func>
+auto wrap_pybind_function(Func&& f) {
```

Comment:
This is the same as `_get_safe_version` but should work with any callable and doesn't require an extra `std::function` layer.